### PR TITLE
xbutil validate with a VCU soft kernel test case / example

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
Added a xbutil validate test case for the VCU softkernel validation for **decoder**. 
The test application will run with XRT API.
Please find the details of the application:
1.	Sending a single frame as an input buffer array.
2.	Input buffer is hardcoded into a header file as a hex array
3.	Expected(Golden) output is also saved on the header file. 
4.	Hard-coded gstreamer specific filed i.e. video information based on the input frame we are using for this test. 
5.	Verify the output for correctness
6.	Code is placed under XRT/tests/xrt/vcu_test directory.

### CODE REVIEW

**Comment by xilinxgitops** [comment]
retest this please

File: tests/validate/xcl_vcu_test/src/plugin_dec.cpp [hunk 1]
```diff
@@ -0,0 +1,1141 @@
...
+  memset (payload_buf, 0, priv->sk_payload_buf->size);
+
+  payload_buf->cmd_id = VCU_FLUSH;
+  iret =
+      xclSyncBO (priv->xcl_handle, priv->sk_payload_buf->bo,
+      XCL_BO_SYNC_BO_TO_DEVICE, priv->sk_payload_buf->size, 0);
+  if (iret != 0) {
+    ERROR_PRINT ("synbo failed - %d, reason : %s", iret,
+        strerror (errno));
+    goto error;
```
Comment (AShivangi) [inline tests/validate/xcl_vcu_test/src/plugin_dec.cpp:511]: same as above

**Review by sarab96 — APPROVED** [review]
Looks good.
Changes may be required for xclbin location/name, etc and for validate requirements.
Looks complex for a validate test :)

**Review by AShivangi — APPROVED** [review]
Minor comments about coding

**Comment by saifuddin-xilinx** [comment]
> Looks good.
> Changes may be required for xclbin location/name, etc and for validate requirements.
> Looks complex for a validate test :)



> Looks good.
> Changes may be required for xclbin location/name, etc and for validate requirements.
> Looks complex for a validate test :)

I have to write the almost the full plugin to verify decoder. 
To verify decoder the vcu kernel need to configure before send the actual frame.

File: tests/validate/xcl_vcu_test/src/host.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
+/**
+* Copyright (C) 2020 Xilinx, Inc
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply tests/validate/xcl_vcu_test/src/host.cpp]: Updated

File: tests/validate/xcl_vcu_test/src/plugin_dec.cpp [hunk 3]
```diff
@@ -0,0 +1,1141 @@
...
+  payload_buf = (sk_payload_data *) (priv->sk_payload_buf->user_ptr);
+  payload_buf->cmd_id = VCU_INIT;
+  payload_buf->obuff_num = priv->num_out_bufs;
+
+  iret = xclSyncBO (priv->xcl_handle, priv->sk_payload_buf->bo,
+      XCL_BO_SYNC_BO_TO_DEVICE, priv->sk_payload_buf->size, 0);
+  if (iret != 0) {
+    ERROR_PRINT ("syncbo failed - %d, reason : %s", iret,
+        strerror (errno));
+    goto error;
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply tests/validate/xcl_vcu_test/src/plugin_dec.cpp]: Resolved all the required places.

File: tests/validate/xcl_vcu_test/src/plugin_dec.cpp (hunk 1 above)
Comment (saifuddin-xilinx) [reply tests/validate/xcl_vcu_test/src/plugin_dec.cpp]: Resolved all the required places.

**Review by stsoe — COMMENTED** [review]
The description of this PR says that it is using XRT APIs, yet I see OpenCL and XCL, and not XRT.  I am a little confused as to what I should be reviewed and how?  Is this C++ code, is it supposed to be an example customers will refer to?  Is for internal use?    Let me know so I can review approrpiately.

**Comment by AShivangi** [comment]
@stsoe, This is a test case that will be used by the new xbutil validate flow. We are adding this test case, but hopefully, we will be able to transfer ownership to Heera's team in the future as they will be owning and maintaining all the xbutil validate test cases

File: tests/validate/xcl_vcu_test/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  set(OpenCL_INCLUDE_DIR ${OCL_ROOT}/include)
+  find_library(OpenCL_LIBRARY
+    NAMES OpenCL
+    HINTS "${OCL_ROOT}/lib")
+  include_directories(${OpenCL_INCLUDE_DIR})
+else()
+  find_package(OpenCL)
+endif(WIN32)
+
+include_directories(../../../src/include/1_2 ../../../src/runtime_src/core/include src/ ../common/includes/xcl2 ../common/includes/cmdparser ../common/includes/logger )
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply tests/validate/xcl_vcu_test/CMakeLists.txt]: Removed src

**Review by chvamshi-xilinx — APPROVED** [review]

**Comment by chvamshi-xilinx** [comment]
@stsoe , this is needed to enable VCU test in xbuitll validate for U30. We have taken utilities from Gstreamer team and created a standalone(xcl based) test.
Lets merge this change and we can cleanup after U30 Beta Release

**Review by stsoe — APPROVED** [review]

**Pull request merged**